### PR TITLE
Create data storage and access API

### PR DESCRIPTION
**Pull request by scott97**
Resolves issue #4 

This adds the following API urls for fetching data: /api/v1/capturesources
/api/v1/videostreams
/api/v1/annotations

The API handlers are placeholders and return 0 items since the datastores are yet to be created.

### CODE REVIEW

**Comment by scott97** [comment]
Marked as draft as files are missing the author + licence comment

File: api/handlers/annotations.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	From time.Time `json:"from"`
+	To   time.Time `json:"to"`
+}
+
+type BoundingBox struct {
+	x1, x2, y1, y2 int
+}
+
+type AnnotationResult struct {
+	Id            *int               `json:"id,omitempty"`
```
Comment (scruzin) [inline api/handlers/annotations.go:54]: The Go idiom is for abbreviations to be fully capitalized, so it is ID not Id.

File: api/handlers/annotations.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	VideoStreamId *int               `json:"videostreamId,omitempty"`
```
Comment (scruzin) [inline api/handlers/annotations.go:55]: Ditto. VideoStreamID not VideoStreamId.

File: api/handlers/capturesource.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import (
+	"fmt"
+
+	"api/utils"
+
+	"github.com/gofiber/fiber/v2"
+)
+
+type CaptureSourceResult struct {
+	Id       *int    `json:"id,omitempty"`
```
Comment (scruzin) [inline api/handlers/capturesource.go:45]: Id -> ID

File: api/handlers/videostream.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+import (
+	"api/utils"
+	"fmt"
+	"time"
+
+	"github.com/gofiber/fiber/v2"
+)
+
+type VideoStreamResult struct {
+	Id        *int           `json:"id,omitempty"`
```
Comment (scruzin) [inline api/handlers/videostream.go:45]: Id -> ID

**Review by trekhopton — COMMENTED** [review]
Let's make the go module directory wide. I believe that is standard practice. You can have multiple packages in a go module, so usually one go module per directory.

Also, don't commit .work and .work.sum files, I think they're part of the local build process and aren't needed. If you use them you might want to add them to .gitignore.

File: api/handlers/annotations.go (hunk 1 above)
Comment (trekhopton) [reply api/handlers/annotations.go:54]: For reference https://github.com/golang/go/wiki/CodeReviewComments#initialisms .

**Comment by trekhopton** [comment]
> Let's make the go module repository wide. I believe that is standard practice. You can have multiple packages in a go module, so usually one go module per repository.
> 
> Also, don't commit .work and .work.sum files, I think they're part of the local build process and aren't needed. If you use them you might want to add them to .gitignore.

That is, have go.mod and go.sum in the root directory of the repo.

File: api/go.mod [hunk 5]
```diff
@@ -0,0 +1,23 @@
+module api
+
+go 1.19
+
+require (
+	github.com/andybalholm/brotli v1.0.5 // indirect
```
Comment (kortschak) [inline api/go.mod:6]: Why are these all indirects?

File: api/handlers/annotations.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+	TimeSpan      *TimeSpan          `json:"timespan,omitempty"`
+	BoundingBox   *BoundingBox       `json:"boundingBox,omitempty"`
+	Observer      *string            `json:"observer,omitempty"`
+	Observation   *map[string]string `json:"observation,omitempty"`
```
Comment (kortschak) [inline api/handlers/annotations.go:59]: No pointer here; `map[T1]T2` is already a pointer under the hood.

File: api/handlers/capturesource.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+package handlers
+
+import (
+	"fmt"
+
+	"api/utils"
```
Comment (kortschak) [inline api/handlers/capturesource.go:39]: Same here.

File: api/utils/result_types.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+package utils
+
+type ApiResult[T any] struct {
```
Comment (kortschak) [inline api/utils/result_types.go:36]: Does this justify the package?

File: go.work
Comment (kortschak) [inline go.work:1]: Delete

File: go.work.sum
Comment (kortschak) [inline go.work.sum:1]: Delete

File: api/utils/result_types.go (hunk 8 above)
Comment (scott97) [reply api/utils/result_types.go:36]: I think so, it currently has code for url parsing and structs for the json responses, and I plan on adding more in the result_types.go file for various API errors as I need

Comment (kortschak) [reply api/utils/result_types.go:36]: It feels to me like these should be in api as `Result` and `Failure`. See https://github.com/golang/go/wiki/CodeReviewComments#package-names.

Similar issue with the functions in the other file in this package.

Comment (scott97) [reply api/utils/result_types.go]: Yes, they are better names

Comment (kortschak) [reply api/utils/result_types.go]: ... but in api, not here.

**Review by trekhopton — APPROVED** [review]

File: api/handlers/annotations.go [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+	To   time.Time `json:"to"`
+}
+
+type BoundingBox struct {
+	x1, x2, y1, y2 int
+}
+
+type AnnotationResult struct {
+	ID            *int              `json:"id,omitempty"`
+	VideoStreamID *int              `json:"videostreamId,omitempty"`
```
Comment (saxon-milton) [inline api/handlers/annotations.go:56]: What's the rationale behind making these pointers ?

File: api/handlers/annotations.go [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+	TimeSpan      *TimeSpan         `json:"timespan,omitempty"`
+	BoundingBox   *BoundingBox      `json:"boundingBox,omitempty"`
+	Observer      *string           `json:"observer,omitempty"`
+	Observation   map[string]string `json:"observation,omitempty"`
+}
+
+func GetAnnotationByID(ctx *fiber.Ctx) error {
+	// TODO: implement handler
```
Comment (saxon-milton) [inline api/handlers/annotations.go:64]: It would be nice to link todos with issues in future.

**Pull request merged**